### DISCUSSION TITLE
How to use "generateStaticParams" with NextJS "custom request handlers" (app/api/params/route.js)?

### DISCUSSION THREAD

**Discussion by vaibhav17sharma** [vercel/next.js#66255] [App Router]
### Summary

i'm facing issue in using this functionality as i'm try to achieve "generateStaticParams" feature, here is the piece of code in which i'm getting issue while making next build.
![image](https://github.com/vercel/next.js/assets/64068490/611256aa-cd81-42fa-af70-f3ade589bcb8)

the issue screenshot is provided here.
![image](https://github.com/vercel/next.js/assets/64068490/7dfad55b-8c21-425e-802a-58c38f8eb552)

on the other hand when i'm using same thing with using JSON data from project directory instead of API route, it's working fine. Please help how can i achieve this using the NExt js Full stack Capabilities.

**Comment by serdar-webdev** [comment]
I saw you use APP router so there is no any need to use generateStaticParams, because of app router works on the server by default, so you can just put this logic into page.jsx, if it won't work, let me know!

**Reply by vaibhav17sharma** [reply]
yes App router works on the server by default, but i',m having a flow where slug needs to be api driven based on the data coming from the api it maybe the API router issue because i'm accessing the database inside the API routes of the projects.

**Comment by serdar-webdev** [comment]
And if we will follow the best practises, in best case use server actions + fetchingData on the server; instead of using api handler, i know here can be situations while you need to use api handler, then use it.

**Comment by hyjoong** [comment]
Is there a specific reason you are using API routes to fetch slug data? Did you write code to access the database directly in the API routes? I'm not sure how the slugData() communication is implemented, and I'm curious why it is necessary to use API routes for fetching slug data.

As a precaution, try declaring 'use server' at the top of the file where slugData() is used.

The build error might be occurring because the slugData() communication is failing, resulting in data not being an array. This could lead to an error when attempting to use the map() method. To handle potential errors during data fetching, it would be a good idea to implement a try-catch block.

```javascript
export async function generateStaticParams() {
  let response = [];
  try {
    let data = await slugData();
    data.forEach((item) => {
      response.push({ slug: item.slug });
    });
  } catch (error) {
    console.error("Error fetching slug data:", error);
  }
  return response;
}
```

**Reply by vaibhav17sharma** [reply]
Yes, i'm accessing the Database directly in the API routes, this is how my api end point returns the data when app is running in development mode.
![image](https://github.com/vercel/next.js/assets/64068490/2f5b43d6-03f5-4dbf-aacc-f5697a2e06ab)
and this is the definition of  slugData function 
![image](https://github.com/vercel/next.js/assets/64068490/ec9bdad2-59bc-4c87-b2cc-128c4a071e03)

for the blog flow i used static data from project directory and it work fine but for api returned slug its creating issue. I'm not sure as it is standard way of using it as, currently i'm learning with experimenting different approaches. maybe Using API routes to get data is the issue here. here are the static routes that are generated for blogs which using static data stored as JSON in project directory
![image](https://github.com/vercel/next.js/assets/64068490/bb8a57d3-8b47-4560-a474-ed636d0a11a2)

**Comment by vinnymac** [comment]
It is described how to do this here: https://github.com/vercel/next.js/issues/51147#issuecomment-1842197049

Although it has the massive caveat of not working out of the box and requiring you to configure `output: 'export'` which is very restricting, and basically a hack to workaround the fact this doesn't **Just Work™**

Related to: https://github.com/vercel/next.js/discussions/55890
